Base functionality, sphinx and PyPi config (#3)
AmeerHamoodi/EEGPhasePy: docs/source/usage/etp.rst

ETP-based phase estimation
===========================

Background
------------
Educated Temporal Prediction (ETP) is a form of phase estimation that uses the time of the latest peak in the current time window and the average
inter-peak-interval to predict the next time the desired phase will occur at. ETP was first described by :cite:t:`Shirinpour2020-ef`, so please cite

> :cite:t:`Shirinpour2020-ef` when using the ETP class of EEGPhasePy.

Reviewer: ```suggestion
:cite:t:`Shirinpour2020-ef` when using the :class:`ETP` class of EEGPhasePy.
```
In RST, use [cross-refs](https://www.sphinx-doc.org/en/master/usage/domains/python.html#cross-referencing-python-objects) or use double-\` for code snippets(\``code``)

Author: Sounds good, updating code with cross-refs and code snippets

The primary benefit of ETP over other phase estimation methods is its ability to work in environments with large and uncertain (within limits) 
delays. This is especially useful considering most EEG amplifiers don't make gaurentees about communication delays and the computer running ETP in real-time will likely suffer from

> jitter in scheduling your phase triggered stimulus (e.g. tES or TMS). ETP allows you to account for these delays while suffering less of an accuracy

Reviewer: Maybe explain acronyms (tES, TMS), or make them into links to wikipedia?

Author: Good point, updating

> dip compared to other phase estimation algorithms

Reviewer: Is there a reference for this?

Author: There isn't a reference showing this, but to the best of my knowledge, all other forms of phase estimation can only accurately predict the phase at the current point in time. So any delay or jitter really impacts the accuracy of phase estimation. ETP provides the time that your target phase will next occur at so, you can schedule triggers rather than detect phase, it can account for delay and sort of help with jitter (you can better identify where the jitter is coming from with ETP and deal with the source of the jitter). With the estimators that predict the current phase, you could potentially forward forecast the time till your target phase using an estimate of the current peak frequency or mean IPI, but at that point the algorithm becomes a more computationally expensive ETP

In short, those claims are an inference made based on the literature rather than something that has been empirically tested

Usage
---------

> Using ETP is quite simple. We'll start off by importing the ETP class, numpy to simulate EEG data and scipy.signal for filtering.

Reviewer: ```suggestion
Using ETP is quite simple. We'll start off by importing the ETP class, :mod:`numpy` to simulate EEG data and :mod:`scipy.signal` for filtering.
```
Using inter sphinx mappings you can automatically produce clickable links like this. Just add inter sphinx to `conf.py` like:
 - https://github.com/Eelbrain/Eelbrain/blob/main/doc/conf.py#L41
 - https://github.com/Eelbrain/Eelbrain/blob/main/doc/conf.py#L239

.. code-block:: Python
   :caption: Imports
   import numpy as np
   import scipy.signal as signal

   from EEGPhasePy.estimators import ETP

> Next, we'll create 2 200s long signals simulating the human alpha rhythym (assuming 10hz here) and adding in some gaussian noise.

Reviewer: ```suggestion
Next, we'll create two 200 s long signals simulating the human alpha rhythym (assuming 10 Hz here) and adding in some gaussian noise.
```

Our training signal will be used to fit the ETP algorithm and then we will test it on the testing signal.

> .. code-block:: Python

Reviewer: Consider [doctest](https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html) for code blocks

Author: Added doctest for most code blocks

>    from EEGPhasePy.estimators import ETP

Reviewer 2: This could not be recognized in Python 3.10.
Using Python 3.11 allowed this to work.

Author: Removed typing issue that caused this issue in 3.10 and below. EEGPhasePy should be compatible with Python 3.9 and above. Thinking of also adding a job for each major Python version supported. Looks like MNE does something similar https://github.com/mne-tools/mne-python/blob/main/.github/workflows/tests.yml, will update our GH actions to do this as well

.. code-block:: Python
   :caption: Imports

   import numpy as np
   import scipy.signal as signal

   from EEGPhasePy.estimators import ETP

Next, we'll create two 200 s long signals simulating the human alpha rhythym (assuming 10 Hz here) and adding in some gaussian noise.

> Our training signal will be used to fit the ETP algorithm and then we will test it on the testing signal.

Reviewer 2: In addition to generating training data, would it also be beneficial to include a section where we can import a publicly available dataset, to then apply this etp-approach onto? This way, it can allow users to understand how to import their own datafiles

Author: Added section using real data

After our signals have been created, we need to construct our real-time and ground-truth filter. For simplicity, 
we have assumed that a higher order filter will allow us to effectively obtain ground truth data for most of the signal. 
The purpose of the ground-truth filter is to obtain the true phase. This filter would be used when training ETP and for 
computing phase stats (e.g. circular mean and std) based on triggers. The real-time filter is used to filter the raw window

> of EEG data passed into the `predict` method.

Reviewer: ```suggestion
of EEG data passed into the :meth:`ETP.predict` method.
```

.. note:: A note on the "ground truth" filter 

   We use ground-truth here quite loosely. It is very challenging to
   obtain a true "ground-truth" for EEG phase. If you're interested in understanding
   more of the nuance associated with this, check out :cite:t:`Zrenner2020-zb`.

.. code-block:: Python
   :caption: Constructing our filters
   
   rt_filter = signal.firwin(120, [8, 12], fs=fs, pass_zero=False)
   gt_filter = signal.firwin(300, [8, 12], fs=fs, pass_zero=False)

Next, we will instantiate ETP and fit it to our training data. The `min_ipi` parameter specifies the shortest
time between peaks the algorithm should expect. This is used to limit the effect phase-slips or phase-resets have
on fitting ETP.

.. code-block:: Python
   :caption: Constructing and fitting ETP
   
   etp = ETP(rt_filter, gt_filter, fs)
   etp.fit(training_signal, min_ipi=int(fs * 1/12))

> Then, we will run a psuedo-real-time simulation using the testing signal. Here, we use the `predict` method of the `etp`

Reviewer: ```suggestion
Then, we will run a pseudo-real-time simulation using the testing signal. Here, we use the `predict` method of the `etp`
```

object. `predict` works by returning the next sample your target phase will occur at.

.. code-block:: Python

>    :caption: Psuedo-real-time simulation with ETP

Reviewer: ```suggestion
   :caption: Pseudo-real-time simulation with ETP
```

>    window_i = 0

Reviewer: Could call it `window_start`? (`i` is often associated with counting, i.e., `for i in windows: ...`)

Author: That naming makes more sense, updated

   etp.fit(training_signal, min_ipi=int(fs * 1/12))

Then, we will run a pseudo-real-time simulation using the testing signal. Here, we use the `predict` method of the `etp`
object. `predict` works by returning the next sample your target phase will occur at.

.. testcode::

   window_start = 0
   window_len = int(0.5*fs)
   window_step = int(0.06*fs) # using a step of 60 ms

   triggers = []

   while window_start + window_len < len(testing_signal):
       window = testing_signal[window_start:window_start + window_len]
       

>        if window_start + window_len + etp.predict(window, 0) < len(testing_signal) - window_len:

Reviewer 2: In Python 3.10, etp.predict(window, 0) gave the error: raise TypeError: Value must be a float type. Fix was to change to etp.predict(window, 0.0) 
In Python 3.11 however, this error did not occur

Author: Updated to allow both float and int

.. nbgallery::

   ../examples/etp-based-phase-estimation

> References

Reviewer: There's something weird going on with the generated bibliography. Maybe you can have a single bibliography page for the whole docs?

Author: Created one bibliography page for the whole docs

>    waveform_fig = etp.plot_mean_std_waveform_from_triggers(testing_signal, triggers, tmin=0.1, tmax=0.1)

Reviewer 2: It would be helpful to display the output figures on the tutorial itself, followed by a brief interpretation of each figure for the reader to understand)

Author: Figures now displayed in tutorial

Examples 
----------

> .. minigallery:: ../examples/etp-*

Reviewer 2: Thumbnail image could depict the output figure of this example.

Author: Added

> .. minigallery:: ../examples/etp-*

Reviewer 2: Also, it may be more organized if the 'Alpha phase estimation using ETP' on the sidebar was listed directly beneath  'ETP-based phase estimation' as a subsection/subheading

Author: good idea, updated
